Refactor Shooter Arm Angle Code

awfulcarpet opened the pull request:
Resolves #20  

Moves Shooter Arm Motor position changes into setAngle method which is called by Amp and Speaker Commands

(Should I move the base of my branch from Hailey's to Main as the Wrist Calibration hasn't been merged into the main branch?)

Stubbynarwhal committed:
Wrist CANcoder Calibration

awfulcarpet committed:
add setAngle abstraction to Shooter subsystem

allows for arbitrary values to be set for pivot motor
allows for less static changing of position of arm

currently not called by anything

awfulcarpet committed:
add telemetry for targetArmAngle

add telemetry to moniter angles set by setAngle

awfulcarpet committed:
apply setAngle method on all commands

awfulcarpet committed:
fix canshoot method in Shooter Subsystem

before we were checking if the motors were still running to end check
the conditions.  however, if setControl continuously sets the motors, it
will always be running.

Instead, check whether the current position of the motor is within a
certain error threashold.

awfulcarpet committed:
actually use pivotMotorAngleErrorThreashold

I forgot to save the file and missed it when reviewing the commit

Jnesselr commented:
> (Should I move the base of my branch from Hailey's to Main as the Wrist Calibration hasn't been merged into the main branch?)

That depends. If your code requires hers to work, then it should be based on her branch. She could make a PR for just that portion and we could land it. The other option that works fine here is just treating it as the one unit of work and either you or @Stubbynarwhal push to get this merged in.

Jnesselr commented on src/main/java/frc/robot/commands/AmpCommand.java:
```diff
@@ -21,7 +20,8 @@ public AmpCommand(Shooter shooter) {
   @Override
   public void execute() {
-    shooter.startMotors(speed, Position.AMP);
+    shooter.setAngle(36);
```
I know this is a magic value right now, but I'd still prefer it to be in a constant field or something with a TODO stating that it needs to be tuned. You will also want to specify once you tune it where on the field the robot needs to be for that angle to be correct.

Jnesselr commented on src/main/java/frc/robot/commands/calibration/CalibrateWristAngleCommand.java:
```diff
@@ -0,0 +1,49 @@
+        double wristEncoderOffsetAtZero = currentProgrammedOffset + wristOffsetWithCurrentProgrammedOffset;
+        SmartDashboard.putNumber("New Wrist Zero offset", wristEncoderOffsetAtZero);
+    }
+}
```
The little circle with a minus sign in the middle means there's no newline at the end of this file. You should always have the newline character at the end of the file. I don't mean an empty line, but the last line of the file should always end in a newline. It's a standard across the industry, more or less.

Jnesselr commented on src/main/java/frc/robot/generated/TunerConstants.java:
```diff
@@ -14,6 +14,8 @@
 
 public class TunerConstants {
+    // constants for pivotMotor for shooter
+    public static final double pivotMotorAngleErrorThreashhold = 1.0 / 360.0;
```
`TunerConstants` are technically for swerve. I'd ideally like to keep it just swerve constants and maybe rename this class if it's confusing. I also don't think all constants need to be in this one location. There's a lot of pros and cons for either approach of putting everything in one constants file vs putting them in the class they relate to. I tend to lean towards the latter approach.

Jnesselr commented on src/main/java/frc/robot/subsystems/Shooter.java:
```diff
@@ -4,74 +4,85 @@
-      // motionMagicVoltage = new MotionMagicVoltage(0.1);  // TODO: Change to fit speaker angle -- not controled yet
-    }
-    positionOfArm = Position.DEFAULT;
+    setAngle(0);
```
Our angle won't be exactly 0 degrees. There's also the question of whether our default means "resting down" or "resting up". I think resting down is ~30 degrees and is probably fine for driving around with, especially because we need it there to load a new note anyway. The name `setDefault` might need to be specific as to what we consider the default state to be, but that's just a little nitpicky.

Jnesselr commented on src/main/java/frc/robot/subsystems/Shooter.java:
```diff
@@ -4,74 +4,85 @@
 
-  public void startMotors(double speed, Position position) {
+  public void setAngle(double angle) {
+    motionMagicVoltage = new MotionMagicVoltage(angle / 360);
+    motionMagicVoltage = motionMagicVoltage.withOverrideBrakeDurNeutral(true);
```
Nit, I'd rather have a MotionMagicVoltage field with the fields that don't change (such as the override brake dur neutral value) as a field and then do the with for the value on that field inside the setControl call.

Jnesselr commented on src/main/java/frc/robot/subsystems/Shooter.java:
```diff
@@ -4,74 +4,85 @@
+    pivotConfigs.RotorToSensorRatio = 45 / 8;
+    pivotMotor.getConfigurator().apply(pivotConfigs);
+
+    TalonFX.optimizeBusUtilizationForAll(pivotMotor, leftShootMotor, rightShootMotor);
```
Why are we optimizing can here and inside `optomizeCan` method. I think the `optomizeCan` method should disappear tbh since the order you set up the motors before calling the optimize bus method matters.

Jnesselr commented on src/main/java/frc/robot/subsystems/Shooter.java:
```diff
@@ -4,74 +4,85 @@
+    FeedbackConfigs pivotConfigs = new FeedbackConfigs();
+    pivotConfigs.FeedbackRemoteSensorID = wristPositionEncoder.getDeviceID();
+    pivotConfigs.FeedbackSensorSource = FeedbackSensorSourceValue.RemoteCANcoder;
+    pivotConfigs.RotorToSensorRatio = 45 / 8;
```
You may want to add a comment here explaining your stackup.

Jnesselr commented on src/main/java/frc/robot/subsystems/Shooter.java:
```diff
@@ -4,74 +4,85 @@
+    MagnetSensorConfigs wristPositionMagnetConfigs = new MagnetSensorConfigs();
+    wristPositionMagnetConfigs.AbsoluteSensorRange = AbsoluteSensorRangeValue.Unsigned_0To1;
+    wristPositionMagnetConfigs.SensorDirection = SensorDirectionValue.CounterClockwise_Positive;
+    wristPositionMagnetConfigs.MagnetOffset = 0;  // TODO Calibrate this with a command
```
Since we made the command, I'd specify the command. When it's calibrated, just remove the `TODO` portion so people know how to get a new value if needed.

Jnesselr commented on src/main/java/frc/robot/subsystems/Shooter.java:
```diff
@@ -4,74 +4,85 @@
   private boolean motorsAtShootingSpeed = false;
-  private VelocityVoltage velocity = new VelocityVoltage(targetVelocity);
-  private MotionMagicVoltage motionMagicVoltage = new MotionMagicVoltage(0.1);
+  private MotionMagicVoltage motionMagicVoltage = new MotionMagicVoltage(0);
```
Have you tuned the required Kp, Kd, and Kg parameters yet for this wrist? It's fine if we tune them in Phoenix Tuner X, but I'd like to have them in code too. This is because, if we have to swap a motor out, we can just update it to have the right CAN id and our code will handle updating all of the values to have the correct K values.

Jnesselr commented on src/main/java/frc/robot/subsystems/Shooter.java:
```diff
@@ -4,74 +4,85 @@
 public class Shooter extends SubsystemBase {
 
-  private final TalonFX pivotMotor = new TalonFX(14, Constants.OperatorConstants.canivoreSerial);
+  public final TalonFX pivotMotor = new TalonFX(14, Constants.OperatorConstants.canivoreSerial);
```
I know I suggested this to be public for the `pivotMotor` and the `wristPositionEncoder`, but I still would prefer a getter here so the motor/encoder can't be re-assigned somewhere else. It still lets someone else mess with them which I'm not a huge fan of, but can't think of a cleaner solution at this time.

Jnesselr commented on src/main/java/frc/robot/subsystems/Shooter.java:
```diff
@@ -83,9 +94,11 @@ public void stopMotors() {
   public void periodic() {
     motorVelocity = leftShootMotor.getVelocity().getValue();
     motorsAtShootingSpeed = motorVelocity <= targetVelocity + 10 && motorVelocity >= targetVelocity - 10;
+    positionOfArm = pivotMotor.getPosition().getValue() * 360;
```
Our position isn't based on the motor which is pre-gearbox. It's based on the wrist encoder which is mounted after the gearbox.

Jnesselr commented on src/main/java/frc/robot/subsystems/Shooter.java:
```diff
@@ -83,9 +94,11 @@ public void stopMotors() {
-    SmartDashboard.putString(SMART_DASHBOARD_POSITION, positionOfArm.toString());
+    SmartDashboard.putNumber(SMART_DASHBOARD_POSITION, positionOfArm);
+    SmartDashboard.putNumber(SMART_DASHBOARD_TARGET_POSITION, targetArmPosition);
     pivotMotor.setControl(motionMagicVoltage); // -- not controled yet
```
We should verify this, but we should also remove this call since if you call `setControl` once it will automatically update at 1khz. We shouldn't need to call `setControl` again unless the position we're aiming for changes.

Jnesselr commented on src/main/java/frc/robot/subsystems/Shooter.java:
```diff
@@ -83,9 +94,11 @@ public void stopMotors() {
+    SmartDashboard.putNumber(SMART_DASHBOARD_TARGET_POSITION, targetArmPosition);
     pivotMotor.setControl(motionMagicVoltage); // -- not controled yet
   }
 }
```
Newline at the end of the file here as well.

Jnesselr reviewed (changes requested):
This is an awesome direction. There's only a few things I think are incorrect here, most of this is just improving readability and communicating well with the next developer to look at your code.

awfulcarpet committed:
factor out magic values from Speaker and Amp Command

awfulcarpet committed:
move error constant to shooter.java

move pivotMotorAngleErrorThreashhold to original file for easy tuning
Rename TunerConstants to SwerveTunerConstants to avoid confusion

awfulcarpet committed:
make pivotmotor and encoder private and allow access with getters

awfulcarpet committed:
base positionOfArm on encoder

previous position was based on pre-gearbox, use encoders instead

awfulcarpet committed:
rename setDefault to be more Specific

setdefault was confusing on whatever "default" meant.  instead, rename
it to make it easier to understand that it is getting the arm ready
to recieve a new note from intake

awfulcarpet committed:
change motionMagicVoltage calling

move overrideBrakeDurNeutral to field
call position in setControl

Jnesselr commented on src/main/java/frc/robot/subsystems/Shooter.java:
```diff
@@ -4,74 +4,93 @@
 public class Shooter extends SubsystemBase {
 
-  private final TalonFX pivotMotor = new TalonFX(14, Constants.OperatorConstants.canivoreSerial);
+  private final static TalonFX pivotMotor = new TalonFX(14, Constants.OperatorConstants.canivoreSerial);
```
Making these static isn't really required, we'll only ever have one instance of the Shooter in our robot. Technically someone could instantiate the class and make a second copy, but making this static will only increase the number of weird issues if that happens, not solve them.

Jnesselr reviewed (commented):
@awfulcarpet did you see the comment about the can bus optimization stuff? I added one on not making those motors static as well.

Stubbynarwhal committed:
Removed Static Modifier

awfulcarpet committed:
remove optimizeCan method from shooter subsystem

optimization is already done in the shooterInit() method, no reason to
cal once more

Stubbynarwhal committed:
Removed more statics

Stubbynarwhal committed:
Fixed static again

Stubbynarwhal committed:
Yaya

GucciBanana11 reviewed (approved):
LGTM
